File: Session_07_GraphBasics/multi_language_implementations/graph_search.py
# Session 07: BFS and DFS Traversals in Python
# Uses an Adjacency List graph representation to demonstrate:
# 1. Breadth-First Search (Queue-based)
# 2. Depth-First Search (Recursive)

class Node:
    def __init__(self, dest):
        self.dest = dest
        self.next = None

class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.head_array = [None] * vertices

    # Add undirected edge
    def add_edge(self, src, dest):
        node1 = Node(dest)
        node1.next = self.head_array[src]
        self.head_array[src] = node1

        node2 = Node(src)
        node2.next = self.head_array[dest]
        self.head_array[dest] = node2

# 1. Breadth-First Search (BFS)
def run_bfs(graph, start_vertex):
    visited = [False] * graph.V
    queue = []
    
    visited[start_vertex] = True
    queue.append(start_vertex)
    
    print("BFS: ", end="")
    
    while queue:
        curr = queue.pop(0)
        print(f"{curr} ", end="")
        
        temp = graph.head_array[curr]
        while temp:
            adj = temp.dest
            if not visited[adj]:
                visited[adj] = True
                queue.append(adj)
            temp = temp.next
    print()

# Helper recursive function for DFS
def dfs_helper(graph, vertex, visited):
    visited[vertex] = True
    print(f"{vertex} ", end="")
    
    temp = graph.head_array[vertex]
    while temp:
        adj = temp.dest
        if not visited[adj]:
            dfs_helper(graph, adj, visited)
        temp = temp.next

# 2. Depth-First Search (DFS)
def run_dfs(graph, start_vertex):
    visited = [False] * graph.V
    print("DFS: ", end="")
    dfs_helper(graph, start_vertex, visited)
    print()

def main():
    graph = Graph(5)
    
    # Setup a sample graph:
    # 0 --- 1 --- 3
    # |   /
    # |  /
    # 2 --- 4
    graph.add_edge(0, 1)
    graph.add_edge(0, 2)
    graph.add_edge(1, 2)
    graph.add_edge(1, 3)
    graph.add_edge(2, 4)

    run_bfs(graph, 0)
    run_dfs(graph, 0)

if __name__ == "__main__":
    main()

File: Session_07_GraphBasics/multi_language_implementations/graph_search.rs
use std::io::{self, BufRead};

struct Node {
    dest: usize,
    next: Option<Box<Node>>,
}

struct List {
    head: Option<Box<Node>>,
}

struct Graph {
    v: usize,
    array: Vec<List>,
}

fn create_graph(vertices: usize) -> Graph {
    let mut array = Vec::with_capacity(vertices);
    for _ in 0..vertices {
        array.push(List { head: None });
    }
    Graph { v: vertices, array }
}

fn add_edge(graph: &mut Graph, src: usize, dest: usize) {
    let old_head1 = graph.array[src].head.take();
    graph.array[src].head = Some(Box::new(Node {
        dest,
        next: old_head1,
    }));

    let old_head2 = graph.array[dest].head.take();
    graph.array[dest].head = Some(Box::new(Node {
        dest: src,
        next: old_head2,
    }));
}

fn run_bfs(graph: &Graph, start_vertex: usize) {
    let mut visited = vec![false; graph.v];
    let mut queue = vec![0; graph.v];
    let mut front = 0;
    let mut rear = 0;

    visited[start_vertex] = true;
    queue[rear] = start_vertex;
    rear += 1;

    print!("BFS Starting from Vertex {}: ", start_vertex);

    while front < rear {
        let curr = queue[front];
        front += 1;
        print!("{} ", curr);

        let mut temp = &graph.array[curr].head;
        while let Some(node) = temp {
            let adj = node.dest;
            if !visited[adj] {
                visited[adj] = true;
                queue[rear] = adj;
                rear += 1;
            }
            temp = &node.next;
        }
    }
    println!();
}

fn dfs_helper(graph: &Graph, vertex: usize, visited: &mut Vec<bool>) {
    visited[vertex] = true;
    print!("{} ", vertex);

    let mut temp = &graph.array[vertex].head;
    while let Some(node) = temp {
        let adj = node.dest;
        if !visited[adj] {
            dfs_helper(graph, adj, visited);
        }
        temp = &node.next;
    }
}

fn run_dfs(graph: &Graph, start_vertex: usize) {
    let mut visited = vec![false; graph.v];
    print!("DFS Starting from Vertex {}: ", start_vertex);
    dfs_helper(graph, start_vertex, &mut visited);
    println!();
}

fn main() {
    let num_vertices = 5;
    let mut graph = create_graph(num_vertices);

    add_edge(&mut graph, 0, 1);
    add_edge(&mut graph, 0, 2);
    add_edge(&mut graph, 1, 2);
    add_edge(&mut graph, 1, 3);
    add_edge(&mut graph, 2, 4);

    println!("=== Graph Search Algorithms ===\n");
    run_bfs(&graph, 0);
    run_dfs(&graph, 0);

    println!();
    println!("Press Enter to exit...");
    let mut iterator = io::stdin().lock().lines();
    let _ = iterator.next();
}
